src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AppDirectiveCompletionProviderTests.cs

```diff
@@ -0,0 +1,130 @@
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Completion.CompletionProviders;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests;
+
+public sealed class SdkAppDirectiveCompletionProviderTests : AbstractAppDirectiveCompletionProviderTests
+{
+    protected override string DirectiveKind => "sdk";
+
+    internal override Type GetCompletionProviderType()
+        => typeof(SdkAppDirectiveCompletionProvider);
+}
+
+public sealed class PropertyAppDirectiveCompletionProviderTests : AbstractAppDirectiveCompletionProviderTests
+{
+    protected override string DirectiveKind => "property";
+
+    internal override Type GetCompletionProviderType()
+        => typeof(PropertyAppDirectiveCompletionProvider);
+}
+
+public sealed class PackageAppDirectiveCompletionProviderTests : AbstractAppDirectiveCompletionProviderTests
+{
+    protected override string DirectiveKind => "package";
+
+    internal override Type GetCompletionProviderType()
+        => typeof(PackageAppDirectiveCompletionProvider);
+}
+
+public sealed class ProjectAppDirectiveCompletionProviderTests : AbstractAppDirectiveCompletionProviderTests
+{
+    protected override string DirectiveKind => "project";
+
+    internal override Type GetCompletionProviderType()
+        => typeof(ProjectAppDirectiveCompletionProvider);
+}
+
+public abstract class AbstractAppDirectiveCompletionProviderTests : AbstractCSharpCompletionProviderTests
+{
+    protected abstract string DirectiveKind { get; }
+
+    protected static string GetMarkup(string code, string features = "FileBasedProgram=true") => $$"""
+        <Workspace>
+            <Project Language="C#" CommonReferences="true" AssemblyName="Test1" Features="{{features}}">
+            <Document><![CDATA[{{code}}]]></Document>
+            </Project>
+        </Workspace>
+        """;
+
+    [Fact]
+    public Task AfterHashColon()
+        => VerifyItemExistsAsync(GetMarkup("""
+            #:$$
+            """), expectedItem: DirectiveKind);
+
+    [Fact]
+    public Task AfterHashColonQuote()
+        => VerifyItemIsAbsentAsync(GetMarkup("""
+            #:"$$
+            """), expectedItem: DirectiveKind);
+
+    [Fact]
+    public Task NotWhenFileBasedProgramIsDisabled()
+        => VerifyItemIsAbsentAsync(GetMarkup("""
+            #:$$
+            """, features: ""), expectedItem: DirectiveKind);
+
+    [Fact]
+    public Task AfterHashColonSpace()
+        => VerifyItemExistsAsync(GetMarkup("""
+            #: $$
+            """), expectedItem: DirectiveKind);
+
+    [Fact]
+    public Task NotAfterHashColonSpaceColon()
+        => VerifyItemIsAbsentAsync(GetMarkup("""
+            #: :$$
+            """), expectedItem: DirectiveKind);
+
+    [Fact]
+    public Task NotAfterHashColonWord()
+        => VerifyItemIsAbsentAsync(GetMarkup("""
+            #:word$$
```

Comment:
Absent seems right here since `word` isn't a match.  But should add a test for cases like `#: prop$$` to ensure it is present there (provider mentions that is supported).

Reply:
I started to do this manually, but, I found that the VerifyIsPresent helper, actually also verifies that typing out the expected item in the indicated position, continues to recommend the expected item. In other words, if we tried to `VerifyIsPresent("#:prop$$")`, the test would actually fail as it attempts to exercise the `"#:propp$$`, `#:proppr$$`, etc cases.